src/object.d

```diff
@@ -3287,135 +3302,57 @@ int __cmp(T1, T2)(T1[] s1, T2[] s2)
         }
         return s1.length < s2.length ? -1 : (s1.length > s2.length);
     }
-    // floating point types
-    else static if (__traits(isFloating, U1))
+    // Imaginary types
+    else static if (is(U1 == ifloat) || is(U1 == idouble) || is(U1 == ireal))
     {
+        // Special-case imaginary types to use comparison for "concrete" types
         static if (is(U1 == ifloat)) alias F = float;
         else static if (is(U1 == idouble)) alias F = double;
         else static if (is(U1 == ireal)) alias F = real;
```

Comment:
Let's say we have:

```d
int[] a;
const(int)[] b;
immutable(int)[] c;

writeln(a < b);
writeln(b < c);
writeln(a < c);
```

You said that "the qualifiers are passed by the compiler". I understood that the compiler strips qualifiers before instantiating `__cmp`, so there will be only one `__cmp` instantiation? If so, this should be enforced on the `__cmp` side.

If, however, there are three template instantiations, two of those should short-circuit to calling the third to avoid redundant code gen. I also mentioned this in Lucia's PR and this is what I thought this PR did from your initial description :)